watchtower/wtdb/tower_db_test.go

```diff
@@ -0,0 +1,712 @@
+		updateFromInt(id(0), 2, 1),
+		updateFromInt(id(0), 3, 2),
+		updateFromInt(id(0), 4, 1),
+	},
+	updateErrs: []error{
+		nil, nil, nil, wtdb.ErrLastAppliedReversion,
+	},
+}
+
+func TestTowerDB(t *testing.T) {
+	dbs := []struct {
+		name string
+		init dbInit
+	}{
+		{
+			name: "fresh boltdb",
+			init: func(t *testing.T) (watchtower.DB, func()) {
+				path, err := ioutil.TempDir("", "towerdb")
+				if err != nil {
+					t.Fatalf("unable to make temp dir: %v",
+						err)
+				}
+
+				db, err := wtdb.OpenTowerDB(path)
+				if err != nil {
+					os.RemoveAll(path)
+					t.Fatalf("unable to open db: %v", err)
+				}
+
+				cleanup := func() {
+					db.Close()
+					os.RemoveAll(path)
+				}
+
+				return db, cleanup
+			},
+		},
+		{
+			name: "reopened boltdb",
+			init: func(t *testing.T) (watchtower.DB, func()) {
+				path, err := ioutil.TempDir("", "towerdb")
+				if err != nil {
+					t.Fatalf("unable to make temp dir: %v",
+						err)
+				}
+
+				db, err := wtdb.OpenTowerDB(path)
+				if err != nil {
+					os.RemoveAll(path)
+					t.Fatalf("unable to open db: %v", err)
+				}
+				db.Close()
+
+				// Open the db again, ensuring we test a
+				// different path during open and that all
+				// buckets remain initialized.
+				db, err = wtdb.OpenTowerDB(path)
+				if err != nil {
+					os.RemoveAll(path)
+					t.Fatalf("unable to open db: %v", err)
+				}
+
+				cleanup := func() {
+					db.Close()
+					os.RemoveAll(path)
+				}
+
+				return db, cleanup
+			},
+		},
+		{
+			name: "mock",
+			init: func(t *testing.T) (watchtower.DB, func()) {
+				return wtmock.NewTowerDB(), func() {}
+			},
+		},
+	}
+
+	tests := []struct {
+		name string
+		run  func(*towerDBHarness)
+	}{
+		{
+			name: "create session",
+			run:  testInsertSession,
+		},
+		{
+			name: "delete session",
+			run:  testDeleteSession,
+		},
+		{
+			name: "state update no session",
+			run:  runStateUpdateTest(stateUpdateNoSession),
+		},
+		{
+			name: "state update exhaust session",
+			run:  runStateUpdateTest(stateUpdateExhaustSession),
+		},
+		{
+			name: "state update seqnum equal last applied",
+			run: runStateUpdateTest(
+				stateUpdateSeqNumEqualLastApplied,
+			),
+		},
+		{
+			name: "state update seqnum equal last applied",
+			run: runStateUpdateTest(
+				stateUpdateSeqNumLTLastApplied,
+			),
+		},
+		{
+			name: "state update seqnum zero invalid",
+			run:  runStateUpdateTest(stateUpdateSeqNumZeroInvalid),
+		},
+		{
+			name: "state update skip seqnum",
+			run:  runStateUpdateTest(stateUpdateSkipSeqNum),
+		},
+		{
+			name: "state update revert seqnum",
+			run:  runStateUpdateTest(stateUpdateRevertSeqNum),
+		},
+		{
+			name: "state update revert last applied",
+			run:  runStateUpdateTest(stateUpdateRevertLastApplied),
+		},
+		{
+			name: "multiple breach matches",
+			run:  testMultipleMatches,
+		},
+		{
+			name: "lookout tip",
+			run:  testLookoutTip,
+		},
+	}
+
+	for _, db := range dbs {
+		t.Run(db.name, func(t *testing.T) {
```

Comment:
If you add a `t.Parallel()` here (before the second loop), then all the DB tests will execute in parallel.

Reply:
interesting. adding this surfaced some weird test interactions that only happen when run in parallel, they're fixed in 01948226bf34da79718842e9bbaa17f9e54b9afa